casper/src/main/scala/coop/rchain/casper/BlockDag.scala

```diff
@@ -6,10 +6,26 @@ import coop.rchain.casper.protocol.BlockMessage
 
 import scala.collection.immutable.{HashMap, HashSet}
 
+/*
+ * toposort - Topological sort of the DAG. The index in the outer Vector
+ *            is the block number (length of the longest path back to
+ *            genesis) and the index of the inner Vector breaks ties for
+ *            blocks with the same number based when the node observed the block.
+ *            `topoSort.flatten` gives the sorted blocks.
+ *
+ * sortOffset - This allows us to sort from some block number going forward.
+ *              The idea is that the block height may get too large to use as the index
+ *              of a Vector, or that after a checkpoint on the chain we won't need
+ *              to remember the sorting before that block. The offset gives the block
+ *              number for the first block in the currently maintained sorting.
+ */
```

Comment:
Is this the latest finalized block?

Reply:
No, it's the last checkpoint. Remember from our meeting with @nashef that we are assuming the DAG-related information can be completely stored in-memory between checkpoints.

By the way, `BlockMetadata` contains enough information for the safety oracle I think, so we should update the ticket on refactoring it to make use of this in-memory representation of the DAG.  Then it won't need to hit the database at all and will be super fast again.

Reply:
But if we have `in-memory` representation of the DAG then the memory footprint rises again right?

Reply:
Yes. But holding only the meta-data (i.e. hashes, parent hash pointer and justification hash pointers) it totally doable because it is very little data per block. The main block with it's full proto-representation (including deployed code, logs, etc) will only be on disk and only fetched when needed (which will hopefully be very rarely compared to the metadata usage)